// @vitejs/plugin-rsc/dist/browser.js
import "./dist-BRSdGcl7.js";
import "./shared-DEpnONZf.js";
import { t as setRequireModule } from "./browser-Dtp2JGlz.js";
import { a as createTemporaryReferenceSet, c as setServerCallback, i as createServerReference, n as createFromFetch, o as encodeReply, r as createFromReadableStream, s as findSourceMapURL, t as callServer } from "./browser-BJvYpz0J.js";
import * as clientReferences from "virtual:vite-rsc/client-references";

//#region src/browser.ts
initialize();
function initialize() {
	setRequireModule({ load: async (id) => {
		if (!import.meta.env.__vite_rsc_build__) return __vite_rsc_raw_import__(withTrailingSlash(import.meta.env.BASE_URL) + id.slice(1));
		else {
			const import_ = clientReferences.default[id];
			if (!import_) throw new Error(`client reference not found '${id}'`);
			return import_();
		}
	} });
}
function withTrailingSlash(path) {
	if (path[path.length - 1] !== "/") return `${path}/`;
	return path;
}

//#endregion
export { callServer, createFromFetch, createFromReadableStream, createServerReference, createTemporaryReferenceSet, encodeReply, findSourceMapURL, setRequireModule, setServerCallback };
