// @vitejs/plugin-rsc/dist/rsc.js
import "./dist-BRSdGcl7.js";
import "./encryption-utils-6p8t4Xqm.js";
import "./shared-DEpnONZf.js";
import { a as setRequireModule, i as loadServerAction, r as createServerManifest, t as createClientManifest } from "./rsc-BwsRzIFM.js";
import { a as decodeFormState, c as registerClientReference, i as decodeAction, l as registerServerReference, n as createFromReadableStream, o as decodeReply, r as createTemporaryReferenceSet, s as encodeReply, t as createClientTemporaryReferenceSet, u as renderToReadableStream } from "./rsc-BIM0wUrP.js";
import { n as encryptActionBoundArgs, t as decryptActionBoundArgs } from "./encryption-runtime-DlHpEnUc.js";
import serverReferences from "virtual:vite-rsc/server-references";

//#region src/rsc.tsx
initialize();
function initialize() {
	setRequireModule({ load: async (id) => {
		if (!import.meta.env.__vite_rsc_build__) return import(
			/* @vite-ignore */
			id
);
		else {
			const import_ = serverReferences[id];
			if (!import_) throw new Error(`server reference not found '${id}'`);
			return import_();
		}
	} });
}

//#endregion
export { createClientManifest, createClientTemporaryReferenceSet, createFromReadableStream, createServerManifest, createTemporaryReferenceSet, decodeAction, decodeFormState, decodeReply, decryptActionBoundArgs, encodeReply, encryptActionBoundArgs, loadServerAction, registerClientReference, registerServerReference, renderToReadableStream, setRequireModule };
